src/wasm/wasm-ir-builder.cpp

```diff
@@ -572,6 +605,97 @@ Result<> IRBuilder::visitElse() {
   return Ok{};
 }
 
+Result<> IRBuilder::visitCatch(Name tag) {
+  auto& scope = getScope();
+  bool wasTry = true;
+  auto* tryy = scope.getTry();
+  if (!tryy) {
+    wasTry = false;
+    tryy = scope.getCatch();
+  }
+  if (!tryy) {
+    return Err{"unexpected catch"};
+  }
+  auto originalLabel = scope.getOriginalLabel();
+  auto label = scope.label;
+  auto expr = finishScope();
+  CHECK_ERR(expr);
+  if (wasTry) {
+    tryy->body = *expr;
+  } else {
+    tryy->catchBodies.push_back(*expr);
+  }
+  tryy->catchTags.push_back(tag);
+  pushScope(ScopeCtx::makeCatch(tryy, originalLabel, label));
+  // Push a pop for the exception payload.
+  auto params = wasm.getTag(tag)->sig.params;
+  if (params != Type::none) {
+    push(builder.makePop(params));
+  }
+  return Ok{};
+}
+
+Result<> IRBuilder::visitCatchAll() {
+  auto& scope = getScope();
+  bool wasTry = true;
+  auto* tryy = scope.getTry();
+  if (!tryy) {
+    wasTry = false;
+    tryy = scope.getCatch();
+  }
+  if (!tryy) {
+    return Err{"unexpected catch"};
+  }
+  auto originalLabel = scope.getOriginalLabel();
+  auto label = scope.label;
+  auto expr = finishScope();
+  CHECK_ERR(expr);
+  if (wasTry) {
+    tryy->body = *expr;
+  } else {
+    tryy->catchBodies.push_back(*expr);
+  }
+  pushScope(ScopeCtx::makeCatchAll(tryy, originalLabel, label));
+  return Ok{};
+}
+
+Result<> IRBuilder::visitDelegate(Index label) {
+  auto& scope = getScope();
+  auto* tryy = scope.getTry();
+  if (!tryy) {
+    return Err{"unexpected delegate"};
+  }
+  // In Binaryen IR, delegates can only target try or function scopes directly.
+  // Search upward to find the nearest enclosing try or function scope. Since
+  // the given label is relative the parent scope of the try, start by adjusting
+  // it to be relative to the try scope.
+  ++label;
+  for (size_t size = scopeStack.size(); label < size; ++label) {
+    auto& delegateScope = scopeStack[size - label - 1];
+    if (auto* delegateTry = delegateScope.getTry()) {
+      // Only delegates can reference the try name in Binaryen IR, so trys might
+      // need two labels: one for delegates and one for all other control flow.
+      // These labels must be different to satisfy the Binaryen validator. To
+      // keep this complexity contained within the handling of trys and
+      // delegates, pretend there is just the single normal label and add a
+      // prefix to it to generate the delegate label. The prefix is added on the
+      // try when its scope ends.
```

Comment:
Do we have this complexity in the old parsers?

Reply:
No, because we don't allow branches to anything except for blocks and loops. The old parser also doesn't support delegating to any scope except to a parent `try` directly.

Reply:
I see. Is it not simpler to add a new block as needed, as the old parsers do? (and rely on the optimizer to fold unneeded ones etc.)

But I don't feel strongly here if this feels like the right solution to you. lgtm overall, though I see test failures.

Reply:
We do still add a new block as needed for normal control flow branches that target this `try`. But adding a new block won't help for `delegate`s that target the try because in Binaryen IR, `delegate` must target the `Try` directly and cannot target a block.

Reply:
So what's the difference between the old way and the new way? In the old parser/IR, `try`'s labels can only be targeted from `delegate`s and not branches. Also branches cannot jump to `try`s`; we created an internal block to mimic jumping to a `try`. Are these not true anymore in the new parser/IR? I haven't reviewed the other PRs for the new parser/IRBuilder so I'd appreciate the context.

Reply:
We're not changing the rules for the IR, so it's still the case that `delegate` can only target named `Try` directly in the IR and that normal branches cannot target `Try`. However, in the input text we are parsing with the new parser, we are following the standard and allowing any branch to target a `try` and allowing delegates to target any kind of control flow. Most of the complexity here is because we are converting from input following the standard rules to IR following our more restrictive IR rules on the fly.